Add delay probability, Change bucket definition and fix calcPerc bugs, drawing improvements

ldemailly opened the pull request:
Add delay probability, Change bucket definition and fix calcPerc bugs, drawing improvements

Also adding support for RecordN to record N instance of same data point  
 
Fixes #74

ldemailly committed:
Add support for RecordN to record N instance of same data point

And adding failing test for expected pX

ldemailly commented on stats/stats_test.go:
```diff
@@ -168,6 +169,34 @@ func TestHistogram(t *testing.T) {
 	}
 }
 
+func XTestHistogramData(t *testing.T) {
```
cc @olimpias - this is what I meant/expect
I will add the code now that I have a test :-)

olimpias replied:
Now I see, Thanks for example :).

ldemailly committed:
Fix the cumulative percentile graph

To be flat between gap and draw until the end of the interval and not
the mid point

ldemailly committed:
Adding an extra <=0 bucket and changing bucket interval definition

Most tests pass but there is 1 weird one and some new TODO left before
merging

ldemailly committed:
One more todo

ldemailly committed:
Delay can now also be a list of probabilities

eg
delay="10ms:20,20ms:10,1s:0.5" for 20% 10ms, 10% 20ms, 0.5% 1s and
69.5% 0 delay

ldemailly committed:
Allow optional % at the end of the percentage probability

It helps self documenting the feature

ldemailly committed:
Use variable for offset/multiplier

ldemailly committed:
Url fuzzy escape of % to %25

So it works with ‘invalid’ urls as input like ?delay=100ms:15%
Which should be ?delay=100ms:15%25

Outbound only, curling directly one would still need to send %25 (or
skip trying to send % as they are optional)

ldemailly committed:
Split export and calc percentile(s)

calculate the percentiles after export (scaled data, calculated
cumulative percentage etc…)

The test in this WIP commit are broken because the linear interpolation
isn’t there yet - but I wanted a checkpoint

ldemailly commented:
> The test in this WIP commit are broken because the linear interpolation
isn’t there yet - but I wanted a checkpoint

ldemailly committed:
Added linear interpolation back

ldemailly committed:
Make the linters happy

ldemailly commented:
This is now ready to review. PTAL. TIA

ldemailly commented:
before:
<img width="646" alt="screen shot 2018-01-06 at 1 58 25 pm" src="https://user-images.githubusercontent.com/3664595/34644351-cb210f48-f2e9-11e7-9947-f2504939babd.png">

after:
<img width="646" alt="screen shot 2018-01-06 at 2 01 17 pm" src="https://user-images.githubusercontent.com/3664595/34644367-283ec99a-f2ea-11e7-9268-6641e4358634.png">

douglas-reid commented on fhttp/http.go:
```diff
@@ -57,14 +57,19 @@ var (
 func (h *HTTPOptions) Init(url string) *HTTPOptions {
-	h.URL = url
+	// unescape then rescape % to %25 (so if it was already %25 it stays)
+	h.URL = strings.Replace(strings.Replace(url, "%25", "%", -1), "%", "%25", -1)
```
do we want something here like: `url.QueryEscape()` and `url.QueryUnescape()`  or is the `%` special here?

ldemailly replied:
query escape would mangle / and other characters, % is somewhat special/hacky yes

I'm not sure what to do cleanly here as the url with a raw % is invalid, I'm just making it work despite being invalid so it looks pretty in the UI text field (I think that's also why I didn't have "%" the first time around for the status weights PRs)

douglas-reid commented on fhttp/http_test.go:
```diff
@@ -283,6 +286,38 @@ func TestParseStatus(t *testing.T) {
+		// Good cases
+		{"100ms", 100 * time.Millisecond},
+		{"100ms:100", 100 * time.Millisecond},
+		{"100ms:100%", 100 * time.Millisecond},
```
curious: why allow '%' in the delay string to be valid?

ldemailly replied:
it's the probability percentage - adding the optional % makes it a bit more self documented (I thought, but clearly not ;-) )

douglas-reid replied:
i understood the purpose. i was just wondering if it is worth the hassle.

douglas-reid reviewed (commented):
looks pretty good. just curious about %s in delay specs.

ldemailly committed:
Merge branch 'master' into percentiles-bucket-fix
